lessons/intro/regex/index.md

```diff
@@ -0,0 +1,268 @@
+
+```pycon
+>>> re.match(r"[a-z]{3}[0-9]{2,4}", "cap00096")
+<re.Match object; span=(0, 7), match='cap0009'>
+```
+
+V tomto případě je rovněž nalezena shoda, ale jen na prvních sedmi znacích. Vypořádat se s tímto problémem
+nám pomůže kotva označující konec řádku `$` na konci regulárního výrazu. Tím označíme, že za poslední číslicí už nesmí řetězec pokračovat dále
+a vyřešíme tím náš problém.
+
+```pycon
+>>> re.match(r"[a-z]{3}[0-9]{2,4}$", "cap00096")
+>>>
+```
+
+> [note]
+> Funkce `match` kontroluje shodu s regulárním výrazem od začátku řetězce, a proto nepotřebujeme
+> zadat i `^` jako označení začátku.
+> Funkce `search` naproti tomu hledá shodu kdekoli v řetězci a jak už její název napovídá,
+> hodí se více pro vyhledávání než validaci.
+
+### Vyhledávání
+
+Vyhledávat můžeme buď první část řetězce, která odpovídá regulárnímu výrazu, pomocí `search`, všechny
+odpovídající podřetězce pomocí `findall` nebo jednotlivé shody pomocí `finditer`.
+
+```pycon
+>>> re.search(r".y", "Naše staré hodiny bijí čtyři hodiny.")
+<re.Match object; span=(15, 17), match='ny'>
+>>> re.findall(r".y", "Naše staré hodiny bijí čtyři hodiny.")
+['ny', 'ty', 'ny']
+>>> list(re.finditer(r".y", "Naše staré hodiny bijí čtyři hodiny."))
+[<re.Match object; span=(15, 17), match='ny'>, <re.Match object; span=(24, 26), match='ty'>, <re.Match object; span=(33, 35), match='ny'>]
+```
+
+### Nahrazování
+
+Nahrazení funkcí `sub` funguje velmi podobně jako řetězcová metoda `replace`. Samozřejmě s tím rozdílem, že nahrazená část musí
+odpovídat regulárnímu výrazu.
+
+```pycon
+>>> re.sub(r"\s+and\s+", " & ", "Bread  and butter and    ham  and  spam")
+'Bread & butter & ham & spam'
+>>> re.sub(r"\s+and\s+", " & ", "Bread  and butter and    ham  and  spam", count=1)
+'Bread & butter and    ham  and  spam'
+```
+
+V příkladu výše nahrazujeme slovo `and` obklopené jedním nebo více bílými znaky z obou stran znakem `&` s jednou mezerou na každé straně. Parametrem `count` si můžeme nastavit, kolikrát se má nahrazení provést.
+
+## Nastavení chování (flags)
+
+Jedním ze základních způsobů, jak ovlivnit chování funkcí v modulu `re` jsou tzv. flagy. Jedná se o různá nastavení,
+která je možné v případě potřeby zapnout.
+
+* `re.DEBUG` zobrazí ladící hlášky týkající se zpracování regulárního výrazu
+* `re.IGNORECASE` (zkráceně `re.I`) způsobí, že regulární výraz nebude brát ohled na velikost písmen
+* `re.MULTILINE` (zkráceně `re.M`) umožní zpracování víceřádkových řetězců. `^` a `$` začnou označovat začátek a konec každého řádku místo začátku a konce celého řetězce.
+* `re.DOTALL` (zkráceně `re.S`) způsobí, že se tečka stane zástupným znakem i za nový řádek `\n`, což v základním nastavení není.
+* … a další k nalezení v dokumentaci.
+
+Nejlepší je předávat flagy jako pojmenovaný argument výše zmíněným funkcím. Vyhneme se tím chybám vzniklým kvůli různým počtům pozičních argumentů funkcí v modulu `re`.
```

Comment:
Proč je to nejlepší? Podle mě je nejlepší používat zápis přímo do výrazu, `(?m)`.

Reply:
Tímhle zápisem jsem to nechtěl komplikovat. Měl bych? Já to třeba nikdy nepoužil.

Některé funkce jako `sub` nebo `split` mají ještě číselné argumenty `maxsplit` resp. `count`, takže pokud flags nepředáš jako pojmenovaný, může se stát, že ti to nevyhodí chybu, ale nebude to fungovat dle očekávání, protože to jako číslo polkne jiný poziční argument.

Reply:
Rozumím, co chceš říct. Napsal bych tu větu jinak, ale takhle to taky funguje.